### ISSUE TITLE
MMC_supervised is not deterministic

### ISSUE THREAD

**Issue by hansen7** [metric-learn/metric-learn#179]
#### Description

Hi there, I have found that you have used random sampling in the MMC_supervised, however, I found that by doing this the output transformation matrix is not deterministic, that is, for a fixed input, the fitting results are not identical.

#### Steps/Code to Reproduce
Example:
```python
from metric_learn import MMC_Supervised, MMC
from sklearn.datasets import load_iris
import numpy as np

iris_data = load_iris()
X = iris_data['data']
Y = iris_data['target']

n = X.shape[0]
mask = (Y[None] == Y[:,None])
a, b = np.nonzero(np.triu(mask, k=1))
c, d = np.nonzero(np.triu(~mask, k=1))

for trial_num in range(3):
    mmc = MMC(convergence_threshold=0.01)
    mmc.fit(X, (a, b, c, d))
    L = mmc.transform(np.diag(np.ones(4)))
    print('\n MMC No.%d Trial'%(trial_num+1))
    print(np.dot(L, L.T).round(6))
    
    mmc_s = MMC_Supervised(convergence_threshold=0.01)
    mmc_s.fit(X, Y)
    L = mmc_s.transform(np.diag(np.ones(4)))
    print('\n MMC_Supervised No.%d Trial'%(trial_num+1))
    print(np.dot(L, L.T).round(6))
```

#### Expected Results

The outputs should all be identical(actually this is the exact case in the test code).
```
[[ 0.000514  0.000868 -0.001195 -0.001703]
 [ 0.000868  0.001468 -0.002021 -0.002879]
 [-0.001195 -0.002021  0.002782  0.003964]
 [-0.001703 -0.002879  0.003964  0.005648]]
```

#### Actual Results

```
 MMC No.1 Trial
[[ 0.000514  0.000868 -0.001195 -0.001703]
 [ 0.000868  0.001468 -0.002021 -0.002879]
 [-0.001195 -0.002021  0.002782  0.003964]
 [-0.001703 -0.002879  0.003964  0.005648]]

 MMC_Supervised No.1 Trial
[[ 0.001361  0.000813 -0.00236  -0.002915]
 [ 0.000813  0.000485 -0.001409 -0.001741]
 [-0.00236  -0.001409  0.004092  0.005054]
 [-0.002915 -0.001741  0.005054  0.006243]]

 MMC No.2 Trial
[[ 0.000514  0.000868 -0.001195 -0.001703]
 [ 0.000868  0.001468 -0.002021 -0.002879]
 [-0.001195 -0.002021  0.002782  0.003964]
 [-0.001703 -0.002879  0.003964  0.005648]]

 MMC_Supervised No.2 Trial
[[ 0.000738  0.001115 -0.001666 -0.001599]
 [ 0.001115  0.001683 -0.002516 -0.002413]
 [-0.001666 -0.002516  0.00376   0.003607]
 [-0.001599 -0.002413  0.003607  0.00346 ]]

 MMC No.3 Trial
[[ 0.000514  0.000868 -0.001195 -0.001703]
 [ 0.000868  0.001468 -0.002021 -0.002879]
 [-0.001195 -0.002021  0.002782  0.003964]
 [-0.001703 -0.002879  0.003964  0.005648]]

 MMC_Supervised No.3 Trial
[[ 0.000421  0.000776 -0.00126  -0.001333]
 [ 0.000776  0.00143  -0.002322 -0.002456]
 [-0.00126  -0.002322  0.003769  0.003987]
 [-0.001333 -0.002456  0.003987  0.004217]]
```
for the trails with MMC_surpervised, it is not deterministic.

**Comment by perimosocordiae** [comment]
Yes, this is a consequence of using random sampling to generate constraints from a labeled dataset to approximate supervised learning using a semi-supervised algorithm like MMC.

You can guarantee deterministic behavior for `MMC_Supervised` by passing a fixed `random_state` parameter to the `fit` method.

**Comment by hansen7** [comment]
Is this for the purpose to derive non-trivial results? or to save the computation cost for large data set...?

**Comment by hansen7** [comment]
> Yes, this is a consequence of using random sampling to generate constraints from a labeled dataset to approximate supervised learning using a semi-supervised algorithm like MMC.
> 
> You can guarantee deterministic behavior for `MMC_Supervised` by passing a fixed `random_state` parameter to the `fit` method.

Is this for the purpose to derive non-trivial results? or to save the computation cost for large data set...?

**Comment by wdevazelhes** [comment]
> Is this for the purpose to derive non-trivial results?

What do you mean by non-trivial results ? 

> or to save the computation cost for large data set...?

Yes, if we trained on all possible pairs, we would have `n_samples**2` pairs, which can become huge 

Also, I think having a sampling strategy allows to balance the constraints, i.e. have the same number of positive pairs and negative pairs, which would most of times not be the case if we used all possible pairs

**Comment by hansen7** [comment]
> > What do you mean by non-trivial results ?
> There are some circumstances where the MMC will raise ValueError '...has no non-trivial dissimilarity constraints given for MMC', which might due to the unideal separation property of the input features and similarity constraints. During my implementation, I will return the initial setting of the algorithms, which is np.diag(np.ones(n_features)) :)

**Comment by hansen7** [comment]
>> What do you mean by non-trivial results ?
>
>There are some circumstances where the MMC will raise ValueError '...has no non-trivial dissimilarity constraints given for MMC', which might due to the unideal separation property of the input features and similarity constraints. During my implementation, I will return the initial setting of the algorithms, which is np.diag(np.ones(n_features)) :)

**Comment by hansen7** [comment]
> What do you mean by non-trivial results ?
There are some circumstances where the MMC will raise ValueError '...has no non-trivial dissimilarity constraints given for MMC', which might due to the unideal separation property of the input features and similarity constraints. During my implementation, I will return the initial setting of the algorithms, which is np.diag(np.ones(n_features)) :)

**Comment by hansen7** [comment]
> What do you mean by non-trivial results ?

There are some circumstances where the MMC will raise ValueError '...has no non-trivial dissimilarity constraints given for MMC', which might due to the unideal separation property of the input features and similarity constraints. During my implementation, I will return the initial setting of the algorithms, which is np.diag(np.ones(n_features)) :)

**Comment by wdevazelhes** [comment]
This error should be raised if `MMC` has not managed to find any pair of two similar points or any pair of two dissimilar points, that are actual different points (up to a tolerance of 1e-9), see these lines: 

https://github.com/metric-learn/metric-learn/blob/617f7e5e5d202422f8957dafecea7d361f4110cd/metric_learn/mmc.py#L83-L91

By the way, these lines will be removed in the next release, since we'll assume it's the responsability of the user to have good data. But if the user has regular good data (i.e. data with several classes, and several samples per classes, for instance), it should not raise this error, so it's weird that you see it.

Could you provide a minimum example of when you have this message thrown ? We could investigate whether this is a bug or not

**Comment by hansen7** [comment]
I'am not sure whether this comes from the implicit module bug or the poor of the data. Need to noticing that during my implementation, there is no two identical instances during my training set.


```python
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)
data_dict = dict()

for respondent_id in range(35):
    #print(respondent_id)
    for week_idx in range(1,11,1):
        weighted_pred = weighted_prediction(round_idx=week_idx)
        y_test = weighted_pred[respondent_id].values.reshape(-1, 1)
        mask = (y_test[None] == y_test[:,None])[:,:,0]
        a, b = np.nonzero(np.triu(mask, k=1))
        c, d = np.nonzero(np.triu(~mask, k=1))

        mmc = MMC(convergence_threshold=0.01)
        
        try:
            mmc.fit(X_test.values, (a, b, c, d))
            L = mmc.transform(np.diag(np.ones(9)))
            M = np.dot(L, L.T)
        except ValueError:
            print('R%d_W%d has no non-trivial dissimilarity constraints given for MMC.'%(respondent_id+1, week_idx))
            M = 0.1 * np.diag(np.ones(9))
            
        data_dict['R%d_W%d'%(respondent_id+1, week_idx)] = M    
            
        #print(np.dot(L, L.T).round(2))
        #print(np.dot(L, L.T).round(4) ,"")
        #print('============================================\n')
np.save(r'../new_data/mmc_metric.npy', data_dict)
```
Output:
```
R5_W9 has no non-trivial dissimilarity constraints given for MMC.
```

```python
week_idx = 8
respondent_id = 4

weighted_pred = weighted_prediction(round_idx=week_idx)
y_test = weighted_pred[respondent_id].values.reshape(-1, 1)
mask = (y_test[None] == y_test[:,None])[:,:,0]
a, b = np.nonzero(np.triu(mask, k=1))
c, d = np.nonzero(np.triu(~mask, k=1))
y_test.ravel()
```

Output:
```
array([-5,  6,  6,  6,  5,  6, -5, -5,  3, -4,  6, -5,  6, -5,  6,  6,  6,
       -5, -5, -5, -3,  6, -5,  5,  6,  5, -5,  5,  6,  6, -4,  6, -5,  5,
        6,  5,  5,  6,  6,  6, -5,  6, -3,  5,  5,  6,  5,  5,  5,  6])
```

**Comment by perimosocordiae** [comment]
I think you need to look at `week_idx = 9` to see the case that raised the error.

**Comment by hansen7** [comment]
Oh..I change the week_idx into 9, it seems that it is due to the input data..many thanks

```python
y_test.ravel()
```

```
array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
       -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1])
```
